runtime/lua/vim/lsp/semantic_tokens.lua

```diff
@@ -0,0 +1,233 @@
+	rawset(table, key, rtn)
+	return rtn
+  end})
+local last_tick = {}
+local active_requests = {}
+
+---@private
+local function modifiers_from_number(x, modifiers_table)
+  ---@private
+  local function get_bit(n, k)
+    --todo(theHamsta): remove once `bit` module is available for non-LuaJIT
+    if _G.bit then
+      return _G.bit.band(_G.bit.rshift(n, k), 1)
+    else
+      return (n / math.pow(2, k)) % 2
+    end
+  end
+
+  local modifiers = {}
+  for i = 0, #modifiers_table - 1 do
+    local bit = get_bit(x, i)
+    if bit == 1 then
+      table.insert(modifiers, 1, modifiers_table[i + 1])
+    end
+  end
+
+  return modifiers
+end
+
+M.token_map = {
+  namespace = "LspNamespace",
+  type = "LspType",
+  class = "LspClass",
+  enum = "LspEnum",
+  interface = "LspInterface",
+  struct = "LspStruct",
+  typeParameter = "LspTypeParameter",
+  parameter = "LspParameter",
+  variable = "LspVariable",
+  property = "LspProperty",
+  enumMember = "LspEnumMember",
+  event = "LspEvent",
+  ["function"] = "LspFunction",
+  method = "LspMethod",
+  macro = "LspMacro",
+  keyword = "LspKeyword",
+  modifier = "LspModifier",
+  comment = "LspComment",
+  string = "LspString",
+  number = "LspNumber",
+  regexp = "LspRegexp",
+  operator = "LspOperator"
+}
+
+M.modifiers_map = {
+  declaration = "LspDeclaration",
+  definition = "LspDefinition",
+  readonly = "LspReadonly",
+  static = "LspStatic",
+  deprecated = "LspDeprecated",
+  abstract = "LspAbstract",
+  async = "LspAsync",
+  modification = "LspModification",
+  documentation = "LspDocumentation",
+  defaultLibrary = "LspDefaultLibrary"
+}
+
+---@private
+local function make_cache(base_map)
+  return setmetatable({}, { __index = function(outer_table, ft)
+    local rtn = setmetatable({}, { __index = function(inner_table, key)
+      local hl = base_map[key]
+      if hl then
+	local prefixed
+	if type(hl) == "table" then
+	  prefixed = {}
+	  for k, v in pairs(hl) do
+	    local prefixed_v = ft..v
+	    vim.cmd("highlight default link " .. prefixed_v .. " " .. v)
+	    prefixed[k] = prefixed_v
+	  end
+	else
+	  prefixed = ft..hl
+	  vim.cmd("highlight default link " .. prefixed .. " " .. hl)
+	end
+	rawset(inner_table, key, prefixed)
+	return prefixed
+      end
+      return hl
+    end})
+    rawset(outer_table, ft, rtn)
+    return rtn
+  end})
+end
+
+local token_cache = make_cache(M.token_map)
+local modifiers_cache = make_cache(M.modifiers_map)
+
+---@private
+local function highlight(buf, token, hl)
+  vim.highlight.range(buf, ns, hl, {token.line, token.start_char}, {token.line, token.start_char + token.length})
+end
+
+---@private
+local function highlight_token(buf, token, ft_token_cache, ft_modifiers_cache)
+  local hl = ft_token_cache[token.type]
+  if hl then
+    highlight(buf, token, hl)
+  end
+  for _, m in pairs(token.modifiers) do
+    hl = ft_modifiers_cache[m]
+    -- modifiers can have a per-type mapping
+    -- e.g. readonly = { variable = "ReadOnlyVariable" }
```

Comment:
After using this for while, I wanted to map a combination of TokenType and TokenModifier.

I have to implementation of that, in the first I would just look up modifiers for type.modifier and then for modifier only
```lua
require'vim.lsp.semantic_tokens'.token_map['variable.readonly'] = 'LspReadOnly'
require'vim.lsp.semantic_tokens'.token_map['variable.globalScope'] = 'LspGlobalScope'
```
in the second (latest commit of this PR), you can choose whether you have a direct mapping `readonly = "LspReadOnly"` or whether you want to set a per-type table. This is more efficient since there are only additional look-ups for modifiers that use this feature.
```lua
require'vim.lsp.semantic_tokens'.modifiers_map['readonly'] = { variable = 'LspVariableReadOnly' }
require'vim.lsp.semantic_tokens'.modifiers_map['globalScope'] = { variable = 'LspGlobalScope' }
```